### ISSUE TITLE
suggestion: allow termination argument in argparse to be specified as argument

### ISSUE THREAD

**Issue by 1eed1b13-e6ae-4382-9c4b-f4aa00072df1** [python/cpython#75195]
BPO | [31012](https://bugs.python.org/issue31012)
--- | :---
Nosy | @stevendaprano

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-07-24.13:40:09.842>
labels = ['type-feature', 'library']
title = 'suggestion: allow termination argument in argparse to be specified as argument'
updated_at = <Date 2017-08-04.02:55:15.810>
user = 'https://bugs.python.org/LeonAvery'
```

bugs.python.org fields:
```python
activity = <Date 2017-08-04.02:55:15.810>
actor = 'paul.j3'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-07-24.13:40:09.842>
creator = 'Leon Avery'
dependencies = []
files = []
hgrepos = []
issue_num = 31012
keywords = []
message_count = 5.0
messages = ['298967', '298969', '299609', '299701', '299729']
nosy_count = 3.0
nosy_names = ['steven.daprano', 'paul.j3', 'Leon Avery']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31012'
versions = ['Python 3.6']
```

</p></details>

**Comment by 1eed1b13-e6ae-4382-9c4b-f4aa00072df1** [comment]
In the argparse module, the argument '--' is interpreted as a signal that everything after it is a positional argument. '--' is literally written into the argparse code, in \_parse_known_args. This means that a user who wishes to use '--' in some other way is out of luck. 

I suggest the addition of an argument to \_\_init__ (or a method call, or something) that allows '--' to be replaced with some string of the user's choice.

**Comment by stevendaprano** [comment]
> This means that a user who wishes to use '--' in some other way is out of luck. 

I think that's a feature, not a limitation.

Command line arguments should use broadly consistent conventions. I don't want one program to use '--' for "end of options" and another program to use it for "something else".

Suppose you want to use '--' for "foo", and '%%' (say) as the end of options terminator. Whatever foo is. Okay, just swap them -- keep the usual convention where '--' is the options terminator, and use '%%' for foo. That way your users won't be surprised and confused by your application working differently from the great bulk of commandline applications. (Users never read the manual.)

But maybe I'll change my mind if you can show that there's already a very common, and useful, alternative convention of '--' for foo, for some definition of foo.

**Comment by 7a064fe6-c535-4d80-a11f-a04ed39056c5** [comment]
argparse roughly follows POSIX practice:

http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html

    Guideline 10:
    The first -- argument that is not an option-argument should be 
    accepted as a delimiter indicating the end of options. Any 
    following arguments should be treated as operands, even if they 
    begin with the '-' character.

I don't think this kind of thing is fixed, but there should be good reason to make an enhancement.  For many users argparse is already too complicated, and the docs are hard to understand.  Yet another parameter can get lost.

There are some known problems with this '--', specially when there are more than one.  It may also have problems when used with subparers, though I forget the details.

It wouldn't hard to substitute a CONST variable for this '--', and wouldn't require changes to the API.  

Another possibility is to change this '--' to 'parser.prefix_chars[0]*2'.  Documenting that would be more work than implementing it.  And we'd have to add a unittest case.

**Comment by 1eed1b13-e6ae-4382-9c4b-f4aa00072df1** [comment]
The POSIX guideline suggests that -- should be available as an option argument. One should be able to say "command --opt --" in order to make '--' the value of the '--opt' option. 

> There are some known problems with this '--', specially when there are more than one.

Yes. That is exactly the problem I have.

**Comment by 7a064fe6-c535-4d80-a11f-a04ed39056c5** [comment]
The earlier issue for '--' as arguement

http://bugs.python.org/issue14364

With the patch I suggested there '-f--' and '--foo=--' work as arguments.  '--foo --' is still interpreted as a positionals switch.

In optparse parsing, a flagged option gets all the remaining strings, and consume what it needs.  In the argparse parser, strings are classed as flag and argument and '--'.  The top level allocates strings to Actions.  '--' is handled at the top level.  This difference in parsing makes it impossible to complete replicate POSIX action.

See also http://bugs.python.org/issue13922

**Comment by savannahostrowski** [comment]
I don't think we should do this. -- as the end-of-options delimiter is a standard POSIX convention, and the consistency of that across tools is the point, making it customizable would mainly surprise users, for a niche need.
